src/monitor.ts

```diff
@@ -0,0 +1,425 @@
+    this.app.get('/status', (req, res) => {
+      res.send(this.nodesStatus);
+    });
+
+    // Parse all request body as JSON, ignoring content-type, for the /add
+    // route.
+    this.app.use('/add', bodyParser.json({
+      type: '*/*',
+    }));
+
+    // POST /add adds the provided instances to the monitored list. The POST
+    // data must be a JSON object with the following format.
+    // {
+    //   nodes: [IMonitoredNodeConfig...]
+    // }
+    this.app.post('/add', (req, res) => {
+
+      winston.info(req.body);
+
+      if (!req.body.hasOwnProperty('nodes')) {
+        res.status(400).send('There is no field named nodes.');
+        return;
+      }
+
+      if (!(req.body.nodes instanceof Array)) {
+        res.status(400).send('The field nodes is not an array.');
+        return;
+      }
+
+      const validConfigs = [];
+      try {
+        for (const data of req.body.nodes) {
+          const monitoredNodeConfig = validateMonitoredNodeConfig(data);
+          this.nodesStatus.push(this.configToInitialNodeStatus(monitoredNodeConfig));
+          validConfigs.push(monitoredNodeConfig);
+          this.appendConfigData(validConfigs);
+          res.status(200).send('Ok');
+        }
+      } catch (err) {
+        res.status(400).send(`${req.body} contains invalid configuration, error: ${err}`);
+      }
+    });
+  }
+
+  private tryRevive(node: IMonitoredNodeStatus) {
+    if (node.failureCount < this.failureTolerance) {
+      winston.info(`${node.address} hasn't reached failure threshold yet, not reviving...`);
+      return;
+    }
+
+    if (typeof node.reviveCmd === 'undefined') {
+      winston.info(`${node.address} doesn't have a revive command...`);
+      return;
+    }
+
+    let args: string[] = [];
+    if (typeof node.reviveArgs !== 'undefined') {
+      args = node.reviveArgs;
+    }
+
+    const child = ChildProcess.spawn(node.reviveCmd, args, {
+      detached: true,
+      stdio: 'ignore',
+    });
+
+    child.unref();
+  }
+
+  private async ping() {
+    for (const node of this.nodesStatus) {
+      // https://github.com/ethereum/wiki/wiki/JSON-RPC#net_version
+      // We use this simple API as a PING to the network.
+
+      try {
+        const body = await rp.post(`http://${node.address}`, {
+          headers: {
+            'content-type': 'application/json',
+          },
+          json: {
+            jsonrpc: '2.0',
+            method: 'net_version',
+            params: [],
+            // We should ignore any response whose id is smaller than the last
+            // response id to linearize the communication between the monitor
+            // and the network nodes.
+            id: node.lastResponseId + 1,
+          },
+        });
+
+        const nodeResp = this.validateNetVersionResponse(body);
+        if (nodeResp.id > node.lastResponseId) {
+          // Ignore stale responses by checking the request/response id
```

Comment:
in what case will we have stale responses?

Reply:
Consider this, say the heartbeat interval is 3 seconds, we send out request with id 1 at second 0

We send out request id 2 at second 3, we get response for request 2 at second 4

Then we get response for request 1 at second 5, we shouldn't overwrite request 2's response with request 1's response since that's stale. This is possible because the http client library we use might open multiple TCP connections, and everything is performed asynchronously under the hood, one of the connections might get delayed somehow.